Configuration to change breadcrumb uri for child admin

cassianotartari opened the pull request:
<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 3.x is for everything backwards compatible, like patches, features and deprecation notices
    - master is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataAdminBundle/blob/3.x/CONTRIBUTING.md#the-base-branch
-->
I am targetting this branch, because I would like to change default edit route in the breadcrumb for child admins.

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->

Closes #4030

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
-->

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Added
- Configuration to change the default route used to generate the link to the parent object inside a breadcrumb, when in a child admin
- Docs for this configuration

```

## To do

<!--
    If this is a work in progress, COMPLETE and ADD needed tasks.
    You can add as many tasks as you want.
    If some are not relevant, just REMOVE them.
-->

- [x] Update the documentation
- [x] Update code
- [x] Fix existing tests
- [x] Add new tests

## Subject

There is no way to change breadcrumb object link under a child admin. The idea is give a global option to choose another route beyond edit

greg0ire commented on Resources/doc/reference/breadcrumbs.rst:
```diff
@@ -13,3 +13,16 @@ Getting the breadcrumbs for a given action of a given admin is done like this:
    <?php
    $this->get('sonata.admin.breadcrumbs_builder')->getBreadcrumbs($admin, $action);
+
+Child admin default route configuration
```
Just "configuration". As I said, there might be other options later.

greg0ire commented on Resources/doc/reference/breadcrumbs.rst:
```diff
@@ -13,3 +13,16 @@ Getting the breadcrumbs for a given action of a given admin is done like this:
+
+        sonata_admin:
+            breadcrumbs:
+               # use this to change default object link under child admin breadcrumb
```
This is a link to what? The parent, right? If yes, then maybe say `use this to change the default route used to generate the link to the parent object inside a breadcrumb, when in a child admin`

greg0ire commented:
Ok now let's see how the breadcrumbs builder should be called without config. Probably something like `new BreadcrumbsBuilderTest(array('child_admin_route' => 'show'));`
Edit the test file to do just that, and run the test. I believe it should not fail, since php will not complain about an extra argument.
Next, change an expectation, like [this one](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Tests/Admin/BreadcrumbsBuilderTest.php#L350). The test should fail. All you have to do after that is change the breadcrumbs builder to make the test pass.

cassianotartari commented:
I think I messed up everything... Trying to understand the test stuff

greg0ire commented on Admin/BreadcrumbsBuilder.php:
```diff
@@ -82,8 +107,8 @@ public function buildBreadcrumbs(AdminInterface $admin, $action, ItemInterface $
                 array(
-                    'uri' => $admin->hasRoute('edit') && $admin->isGranted('EDIT') ?
-                    $admin->generateUrl('edit', array('id' => $id)) :
+                    'uri' => $admin->hasRoute($this->breadcrumbsConfig['child_admin_route']) && $admin->isGranted(strtoupper($this->breadcrumbsConfig['child_admin_route'])) ?
```
It's not that simple (show requires the View role). See admin::hasAccess IIRC

greg0ire commented on Admin/BreadcrumbsBuilder.php:
```diff
@@ -12,14 +12,39 @@
+    /**
+     * @var string[]
+     */
+    protected $breadcrumbsConfig = array();
```
Just "config" will do

greg0ire commented on DependencyInjection/Configuration.php:
```diff
@@ -73,6 +73,12 @@ public function getConfigTreeBuilder()
+                ->arrayNode('breadcrumbs')
+                    ->addDefaultsIfNotSet()
+                    ->children()
+                        ->scalarNode('child_admin_route')->defaultValue('edit')->end()
```
Use info() to provide some doc here please

greg0ire commented:
It's just style that fails, don't worry and apply THE styleCI is patch

greg0ire commented:
Ok so now let's have a quick look at the build, which consists of many checks. The failing check has a "Details" link. Click on it. You will be presented with a patch showing how to fix styling issues. You can download it and apply it with `git apply patch.txt`, then commit.

Also, you already have four commits, with messages that show trial and errors. We don't really care about that, so let's just squash all commits together `git rebase --interactive origin/3.x`.

This will show you your commits, with actions : 

```
pick aed45 my first commit
pick ded25 my second commit
pick aedf5 my third commit
pick 1ed45 my fourth commit
```

Fixup all commits into the first one : 

```
pick aed45 my first commit
fixup ded25 my second commit
f aedf5 my third commit
f 1ed45 my fourth commit
```
(notice how `fixup` and `f` do the same thing), then `git push --force` to update your PR.

greg0ire commented:
Style is good! Let's look at the unit tests now. Just like you clicked on details for styleci, click Details for Travis CI. You will be presented with a list of jobs. As you can see, jobs about docs and lint are green, but the rest is errr… kinda failing. Let's look at one of them. Click on the line, and look at [the output](https://travis-ci.org/sonata-project/SonataAdminBundle/jobs/148942817#L531).

It says : 

> PHP Fatal error:  Call to a member function getContainer() on a non-object in /home/travis/build/sonata-project/SonataAdminBundle/Admin/AbstractAdmin.php on line 2020

cassianotartari commented:
> Ok so now let's have a quick look at the build, which consists of many checks. The failing check has a "Details" link. Click on it. You will be presented with a patch showing how to fix styling issues. You can download it and apply it with git apply patch.txt, then commit.

First time working with styleci, created an account and I was playing with online interface, I did changes manually and I think now it's working.

> Also, you already have four commits, with messages that show trial and errors. We don't really care about that, so let's just squash all commits together git rebase --interactive origin/3.x.

I did it but I thinks I had to add some message... I just saved it and console thrown:
```bash
Successfully rebased and updated refs/heads/3.x.
```

> Just like you clicked on details for styleci, click Details for Travis CI. You will be presented with a list of jobs. As you can see, jobs about docs and lint are green, but the rest is errr… kinda failing. Let's look at one of them. Click on the line, and look at the output.

Let's try that teacher.

greg0ire commented:
> First time working with styleci, created an account and I was playing with online interface, I did changes manually and I think now it's working.

It is indeed :)

greg0ire commented:
> I did it but I thinks I had to add some message...

Nope, nothing to add, just change all the `pick` into `fixup`, just as advised, then save.

The message is good I think, just `git push --force` so that we can see the result

cassianotartari commented:
I think I don't get it, when I run ``git rebase --interactive origin/3.x`` it opens nano in my console showing this:
```bash
noop

# Rebase 1e61b4f..1e61b4f onto 1e61b4f (1 command(s))
#
# Commands:
# p, pick = use commit
# r, reword = use commit, but edit the commit message
# e, edit = use commit, but stop for amending
# s, squash = use commit, but meld into previous commit
# f, fixup = like "squash", but discard this commit's log message
# x, exec = run command (the rest of the line) using shell
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
# Note that empty commits are commented out
```
I don't get what should write here...

greg0ire commented:
Hum… that's because `origin` is not what I think it is. Can you do a `git remove -vvv` for me ?

cassianotartari commented:
```bash
$ git remote -vvv
origin	https://cassianotartari@github.com/cassianotartari/SonataAdminBundle.git (fetch)
origin	https://cassianotartari@github.com/cassianotartari/SonataAdminBundle.git (push)
```

greg0ire commented:
Oh… you don't even have the main repository as a remote. Let's fix that.

1. Make "origin" the fork : `git remote rename origin fork`
2. Say the main repo is the origin : `git remote add origin git@github.com:sonata-project/SonataAdminBundle.git`
3. Do the rebase, this time it should work
4 `git push -u fork 3.x`

greg0ire commented:
Yup :)

cassianotartari commented:
Good, now the next mission: Travis CI

greg0ire commented:
What about points 3 and 4?

cassianotartari commented:
> What about points 3 and 4?

```
$ git rebase --interactive origin/3.x
fatal: Needed a single revision
invalid upstream origin/3.x
```
:sob::sob::sob:

greg0ire commented:
I forgot : try `git fetch origin` first

cassianotartari commented:
Now works, opened with:
```
pick 51dd43f Configuration to change breadcrumb uri for child admin
pick 829c99d Update breadcrumbs.rst
pick 867f617 Adding the possibility to change the default route used to generate the link to the parent object inside a breadcrumb, when in a child admin. - AbstractAdmin: adding parameter to Breadcrumb co$
pick a5044aa - BreadcrumbsBuilder: changed $breadcrumbsConfig to $config and $admin->isGranted to $admin->hasAccess - Configuration: adding info() to new configuration
pick 1e61b4f styleci
```
Should I edit this before save?

greg0ire commented:
Yes

greg0ire commented:
```
pick 51dd43f Configuration to change breadcrumb uri for child admin
f 829c99d Update breadcrumbs.rst
f 867f617 Adding the possibility to change the default route used to generate the link to the parent object inside a breadcrumb, when in a child admin. - AbstractAdmin: adding parameter to Breadcrumb co$
f a5044aa - BreadcrumbsBuilder: changed $breadcrumbsConfig to $config and $admin->isGranted to $admin->hasAccess - Configuration: adding info() to new configuration
f 1e61b4f styleci
```

cassianotartari commented:
Ok, added this on ``git rebase --interactive origin/3.x`` after I did ``git pull`` and finally:
```$ git push -u fork 3.x
Password for 'https://cassianotartari@github.com': 
Counting objects: 2, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (2/2), done.
Writing objects: 100% (2/2), 542 bytes | 0 bytes/s, done.
Total 2 (delta 0), reused 0 (delta 0)
To https://cassianotartari@github.com/cassianotartari/SonataAdminBundle.git
   1e61b4f..9c1a461  3.x -> 3.x
Branch 3.x set up to track remote branch 3.x from fork.
```

cassianotartari commented:
:shit: 

:cry: I don't know anymore what to do...

greg0ire commented:
Lol you're really not that far. Just don't `git pull`, and don't forget the `--force` option.

cassianotartari commented:
:grin: Done?

greg0ire commented:
Yes!!! Very good!

greg0ire commented:
Next time you need to fix something in a commit, just do : 

- `git commit --amend` (this is like a mini rebase, it fixes the last commit)
- `git push -f`

cassianotartari commented:
:sweat_smile: pretty hot here...

> git commit --amend (this is like a mini rebase, it fixes the last commit)
git push -f

Nice, I'll do this.

Now could we move to Travis CI stuff?

Looking to:
```
1) Sonata\AdminBundle\Tests\Admin\AdminTest::testDefaultBreadcrumbsBuilder
Error: Call to a member function getContainer() on null
```
What can I do to fix it?
```php
$pool = $this->getConfigurationPool();

if($pool) {
    $this->breadcrumbsBuilder = new BreadcrumbsBuilder($pool->getContainer()->getParameter('sonata.admin.configuration.breadcrumbs'));
} else {
    $this->breadcrumbsBuilder = new BreadcrumbsBuilder();
}
```

greg0ire commented:
No, you have to mock the call to `getConfigurationPool()` to return a `Pool` and not nothing, I'm searching how to do that right now.

greg0ire commented:
Fixed! Try this : 

```
        $container = $this->getMock('Symfony\Component\DependencyInjection\ContainerInterface');
        $container->expects($this->once())
            ->method('getParameter')
            ->with('sonata.admin.configuration.breadcrumbs')
            ->will($this->returnValue(array()));

        $pool = $this->getMockBuilder('Sonata\AdminBundle\Admin\Pool')
            ->disableOriginalConstructor()
            ->getMock();
        $pool->expects($this->once())
            ->method('getContainer')
            ->will($this->returnValue($container));

        $admin = $this->getMockForAbstractClass('Sonata\AdminBundle\Admin\AbstractAdmin', array(
            'admin.my_code', 'My\Class', 'MyBundle:ClassAdmin',
        ), '', true, true, true, ['getConfigurationPool']);
        $admin->expects($this->once())
            ->method('getConfigurationPool')
            ->will($this->returnValue($pool));

        $this->assertInstanceOf(
            'Sonata\AdminBundle\Admin\BreadcrumbsBuilder',
            $admin->getBreadcrumbsBuilder()
        );

```

cassianotartari commented:
Wait a second, don't forget that I am a totally noob with tests, where should I add this piece of code?

greg0ire commented:
It's not adding, it's replacing (look at the name of the method, it already exists in `AdminTest`)

greg0ire commented:
Also, please associate your other email address with your account.

cassianotartari commented:
> It's not adding, it's replacing (look at the name of the method, it already exists in AdminTest)

Doing...

> Also, please associate your other email address with your account.

I don't get it, I have 2 emails:
![screenshot from 2016-08-01 15-57-00](https://cloud.githubusercontent.com/assets/3017642/17305404/a04e77ea-5800-11e6-9e7b-e3697ac0f776.png)

greg0ire commented:
What about the one with floripatec ?

cassianotartari commented:
> What about the one with floripatec ?

:grin: I didn't remember where I put this, it's my company email. It could removed, how can I do it?

greg0ire commented:
1. git config --global author.email "your email"
2. Do a git rebase, you have 2 commits instead of one, it should also fix the author.

greg0ire commented:
Sorry, it's `user.email`, not `author.email`. You should edit your ~/.gitconfig and fix this.

cassianotartari commented:
> Sorry, it's user.email, not author.email. You should edit your ~/.gitconfig and fix this.

I did it with ``commit --amend`` and ``push -f`` right now.

greg0ire commented:
Then you should try `git commit --amend --reset-author` I think :P

greg0ire commented:
Looks like it worked ;)

cassianotartari commented:
What is missing now? Wait for travis?

greg0ire commented:
That, and also, add more tests: you just fixed existing tests, but you did not add any. Let's add some, starting with the simplest piece of code : the code inside the configuration class. Find and open `ConfigurationTest.php`, and add a method named `testBreadcrumbsChildRouteDefaultsToEdit`, that tests exactly that.

core23 commented on Admin/BreadcrumbsBuilder.php:
```diff
@@ -21,6 +22,29 @@
+        $this->config = $resolver->resolve($config);
+    }
+
+    public function configureOptions(OptionsResolver $resolver)
```
PHPdoc Missing

cassianotartari replied:
Done.

core23 commented on Admin/AbstractAdmin.php:
```diff
@@ -2017,7 +2017,7 @@ public function buildBreadcrumbs($action, MenuItemInterface $menu = null)
         );
         if ($this->breadcrumbsBuilder === null) {
-            $this->breadcrumbsBuilder = new BreadcrumbsBuilder();
+            $this->breadcrumbsBuilder = new BreadcrumbsBuilder($this->getConfigurationPool()->getContainer()->getParameter('sonata.admin.configuration.breadcrumbs'));
```
This looks very ugly

greg0ire replied:
It is indeed. But will get removed in next major with its surroundings, so, not a real concern IMO

core23 replied:
Do we really need this? This is just a new feature that would work without these ugly hack.

greg0ire replied:
Might be needed by people who overrode the template that contains the breadcrumbs, and still call the breadcrumbs builder the old way. So yeah, it will work for most people if we remove this.

core23 replied:
Can't we switch this? So the breadcrumb builder asks the admin for this property?

greg0ire replied:
The goal since I separated the breadcrumbs builder from the admin is to depend less and less from it, not more. That means that in the end, the breadcrumbs builder should be provided with more and more dependencies through its constructor. Granted, this piece of code is ugly, but will disappear with 4.0, and it's good to have all the ugly code in one place rather than also add a getter in the admin for this parameter, then remove it in 4.0, don't you think?

OskarStark replied:
> Granted, this piece of code is ugly, but will disappear with 4.0, and it's good to have all the ugly code in one place rather than also add a getter in the admin for this parameter, then remove it in 4.0, don't you think?

a PR afterwards which remove this stuff would be great, then

OskarStark replied:
@greg0ire so is there anything todo afterwards in master?

cassianotartari commented:
I'll look all tomorrow morning in my GMT -3, now need to leave. Feel free to edit/add code, if it's possible.

greg0ire commented:
The AdminTest test is ok, now you have to fix the breacrumbs builder test, it seems.

cassianotartari commented:
Hi good morning, let's start.

> That, and also, add more tests: you just fixed existing tests, but you did not add any. Let's add some, starting with the simplest piece of code : the code inside the configuration class. Find and open ConfigurationTest.php, and add a method named testBreadcrumbsChildRouteDefaultsToEdit, that tests exactly that.

I coded this:
```php
    public function testBreadcrumbsChildRouteDefaultsToEdit()
    {
        $config = $this->process(array());
        
        $this->assertSame('edit', $config['breadcrumbs']['child_admin_route']);
    }
```

```bash
$ phpunit Tests/DependencyInjection/ConfigurationTest.php 
PHPUnit 4.7.6 by Sebastian Bergmann and contributors.

.........

Time: 66 ms, Memory: 9.25Mb

OK (9 tests, 17 assertions)
```

greg0ire commented:
Looks good :)

greg0ire commented:
To fix the breadcrumbs builder tests, you have to mock the configuration pool and so on, just like we did in the admin test, and call `$admin->setConfigurationPool($yourPoolMock);`. Can you try that?

cassianotartari commented:
Yes, give me a little time, I'll try to understand the other code and apply it to breadcrumbs tests

cassianotartari commented:
:-1: fail...

Tried to add to ``BreadcrumbsBuilderTest`` this code before each ``getBreadcrumbs`` call
```php
$pool = $this->getMockBuilder('Sonata\AdminBundle\Admin\Pool')
            ->disableOriginalConstructor()
            ->getMock();

$admin->setConfigurationPool($pool);
```

But when I run still with error...
```bash
$ phpunit Tests/Admin/BreadcrumbsBuilderTest.php 
PHPUnit 4.7.6 by Sebastian Bergmann and contributors.

PHP Fatal error:  Call to a member function getContainer() on null in /media/cassiano/cassianosonataadim/SonataAdminBundle/Admin/AbstractAdmin.php on line 2020

Legacy deprecation notices (2)
```

My wisdom tests ended up.

Need to bother you @greg0ire again.

greg0ire commented:
Is it in the same test though? There are several tests.

cassianotartari commented:
> Is it in the same test though? There are several tests.

I'm lost

greg0ire commented:
The `BreadcrumbsBuilderTest` class has several methods, each one is the test. Are you sure the test that fails is the one you fixed? Have a look at the stack trace. Is the failing line still line 182 ?

cassianotartari commented:
So I was looking here https://travis-ci.org/sonata-project/SonataAdminBundle/jobs/148986161#L560

Said this, line 182: ``$admin->getBreadcrumbs('repost');``

So I tried just to add the above poll code before this call, and run ``phpunit`` it throws the same fatal error.

So I go back to travis says about other lines: 182, 262 and 377, tried all too, nothing.

greg0ire commented:
I fetched your branch, tried what you tried (I copy pasted your code on line 66) and… it works! The error message I get is different: 

> Error: Call to a member function getParameter() on null

So now you have to add more mocking (mock the call to getParameter() ) and you will be done for the first test.

cassianotartari commented:
Hell! Pasted the code on line 66 and of ``BreadcrumbsBuilderTest`` and the phpunit error keeps!

```php
    public function testGetBreadcrumbs()
    {
        $class = 'Application\Sonata\NewsBundle\Entity\Post';
        $baseControllerName = 'SonataNewsBundle:PostAdmin';

        $admin = new PostAdmin('sonata.post.admin.post', $class, $baseControllerName);
        $commentAdmin = new CommentAdmin(
            'sonata.post.admin.comment',
            'Application\Sonata\NewsBundle\Entity\Comment',
            'SonataNewsBundle:CommentAdmin'
        );
        $subCommentAdmin = new CommentAdmin(
            'sonata.post.admin.comment',
            'Application\Sonata\NewsBundle\Entity\Comment',
            'SonataNewsBundle:CommentAdmin'
        );
        $admin->addChild($commentAdmin);
        $admin->setRequest(new Request(array('id' => 42)));
        $commentAdmin->setRequest(new Request());
        $commentAdmin->initialize();
        $admin->initialize();
        $commentAdmin->setCurrentChild($subCommentAdmin);

        $menuFactory = $this->getMock('Knp\Menu\FactoryInterface');
        $menu = $this->getMock('Knp\Menu\ItemInterface');
        $translatorStrategy = $this->getMock(
            'Sonata\AdminBundle\Translator\LabelTranslatorStrategyInterface'
        );
        $routeGenerator = $this->getMock(
            'Sonata\AdminBundle\Route\RouteGeneratorInterface'
        );
        $modelManager = $this->getMock(
            'Sonata\AdminBundle\Model\ModelManagerInterface'
        );

        $pool = $this->getMockBuilder('Sonata\AdminBundle\Admin\Pool')
            ->disableOriginalConstructor()
            ->getMock();

        $admin->setConfigurationPool($pool);
        $admin->setMenuFactory($menuFactory);
        $admin->setLabelTranslatorStrategy($translatorStrategy);
        $admin->setRouteGenerator($routeGenerator);
        $admin->setModelManager($modelManager);

        $commentAdmin->setLabelTranslatorStrategy($translatorStrategy);
        $commentAdmin->setRouteGenerator($routeGenerator);

        $modelManager->expects($this->exactly(1))
            ->method('find')
            ->with('Application\Sonata\NewsBundle\Entity\Post', 42)
            ->will($this->returnValue(new DummySubject()));

        $menuFactory->expects($this->exactly(5))
            ->method('createItem')
            ->with('root')
            ->will($this->returnValue($menu));

        $menu->expects($this->once())
            ->method('setUri')
            ->with($this->identicalTo(false));

        $menu->expects($this->exactly(5))
            ->method('getParent')
            ->will($this->returnValue(false));

        $routeGenerator->expects($this->exactly(5))
            ->method('generate')
            ->with('sonata_admin_dashboard')
            ->will($this->returnValue('http://somehost.com'));

        $translatorStrategy->expects($this->exactly(18))
            ->method('getLabel')
            ->withConsecutive(
                array('dashboard'),
                array('Post_list'),
                array('Comment_list'),
                array('Comment_repost'),

                array('dashboard'),
                array('Post_list'),
                array('Comment_list'),
                array('Comment_flag'),

                array('dashboard'),
                array('Post_list'),
                array('Comment_list'),
                array('Comment_edit'),

                array('dashboard'),
                array('Post_list'),
                array('Comment_list'),

                array('dashboard'),
                array('Post_list'),
                array('Comment_list')
            )
            ->will($this->onConsecutiveCalls(
                'someLabel',
                'someOtherLabel',
                'someInterestingLabel',
                'someFancyLabel',

                'someCoolLabel',
                'someTipTopLabel',
                'someFunkyLabel',
                'someAwesomeLabel',

                'someLikeableLabel',
                'someMildlyInterestingLabel',
                'someWTFLabel',
                'someBadLabel',

                'someBoringLabel',
                'someLongLabel',
                'someEndlessLabel',

                'someAlmostThereLabel',
                'someOriginalLabel',
                'someOkayishLabel'
            ));

        $menu->expects($this->exactly(24))
            ->method('addChild')
            ->withConsecutive(
                array('someLabel'),
                array('someOtherLabel'),
                array('dummy subject representation'),
                array('someInterestingLabel'),
                array('someFancyLabel'),

                array('someCoolLabel'),
                array('someTipTopLabel'),
                array('dummy subject representation'),
                array('someFunkyLabel'),
                array('someAwesomeLabel'),

                array('someLikeableLabel'),
                array('someMildlyInterestingLabel'),
                array('dummy subject representation'),
                array('someWTFLabel'),
                array('someBadLabel'),

                array('someBoringLabel'),
                array('someLongLabel'),
                array('dummy subject representation'),
                array('someEndlessLabel'),

                array('someAlmostThereLabel'),
                array('someOriginalLabel'),
                array('dummy subject representation'),
                array('someOkayishLabel'),
                array('dummy subject representation')
            )
            ->will($this->returnValue($menu));

        $admin->getBreadcrumbs('repost');
        $admin->setSubject(new DummySubject());
        $admin->getBreadcrumbs('flag');

        $commentAdmin->getBreadcrumbs('edit');

        $commentAdmin->getBreadcrumbs('list');
        $commentAdmin->setSubject(new DummySubject());
        $commentAdmin->getBreadcrumbs('reply');
    }
```

cassianotartari commented:
Forget! changed the error!

need more coffee

cassianotartari commented:
:confused: 

> add more mocking (mock the call to getParameter() ) and you will be done for the first test.

I was trying to understand how this tests works but when I think I'm getting something I turn back to the first error. I tried to add this before ``setConfigurationPool``
```php
$container = $this->getMock('Symfony\Component\DependencyInjection\ContainerInterface');
$pool->expects($this->any())
    ->method('getContainer')
    ->will($this->returnValue($container));
```
But reading the syntax words seems that I'm not emulating/setting a ``container`` value to ``pool``. I will keep looking other tests to see if I understand somenthing.

greg0ire commented:
> I was trying to understand how this tests works but when I think I'm getting something I turn back to the first error. I tried to add this before setConfigurationPool

Shameless lie! I tried to do exactly that and here is the error message I get : 

> TypeError: Argument 1 passed to Sonata\AdminBundle\Admin\BreadcrumbsBuilder::__construct() must be of the type array, null given, called in /home/users/gparis/dev/SonataAdminBundle/Admin/AbstractAdmin.php on line 2020

So all you have to do after that is mock the call to `getParameter`

cassianotartari commented:
> Shameless lie!

:bowtie:

I was unable to generate this ``construct`` error.

> So all you have to do after that is mock the call to getParameter

Would be same as in ``AdminTest``?
```php
$container->expects($this->once())
    ->method('getParameter')
    ->with('sonata.admin.configuration.breadcrumbs')
    ->will($this->returnValue(array()));
```

Tried to add this entire block on original line 66 and ``Call to a member function getContainer() on null``
```php
        $container = $this->getMock('Symfony\Component\DependencyInjection\ContainerInterface');
        $container->expects($this->once())
            ->method('getParameter')
            ->with('sonata.admin.configuration.breadcrumbs')
            ->will($this->returnValue(array()));
        $pool = $this->getMockBuilder('Sonata\AdminBundle\Admin\Pool')
            ->disableOriginalConstructor()
            ->getMock();
        $pool->expects($this->once())
            ->method('getContainer')
            ->will($this->returnValue($container));

        $admin->setConfigurationPool($pool);
```

greg0ire commented:
> Would be same as in AdminTest?

Yes.

> Tried to add this entire block on original line 66 and Call to a member function getContainer() on null

Are you sure it's for this for `testGetBreadcrumbs`? Maybe you could comment out `testGetBreadcrumbsWithNoCurrentAdmin` ?

cassianotartari commented:
> Are you sure it's for this for testGetBreadcrumbs? Maybe you could comment out testGetBreadcrumbsWithNoCurrentAdmin ?

No. How do I make sure? Is there any parameter to use with ``phpunit`` to show more details?

cassianotartari commented:
News.

Added ``$commentAdmin->setConfigurationPool($pool);`` before ``$commentAdmin->getBreadcrumbs('edit');``

And the code block before ``$admin->getBreadcrumbs('repost');``
```php
$container = $this->getMock('Symfony\Component\DependencyInjection\ContainerInterface');
        $container->expects($this->once())
            ->method('getParameter')
            ->with('sonata.admin.configuration.breadcrumbs')
            ->will($this->returnValue(array()));
        $pool = $this->getMockBuilder('Sonata\AdminBundle\Admin\Pool')
            ->disableOriginalConstructor()
            ->getMock();
        $pool->expects($this->once())
            ->method('getContainer')
            ->will($this->returnValue($container));

        $admin->setConfigurationPool($pool); 
```

And:
```bash
$ phpunit Tests/Admin/BreadcrumbsBuilderTest.php 
PHPUnit 4.7.6 by Sebastian Bergmann and contributors.

F.E....

Time: 146 ms, Memory: 15.25Mb

There was 1 error:

1) Sonata\AdminBundle\Tests\Admin\BreadcrumbsBuilderTest::testUnitChildGetBreadCrumbs
Prophecy\Exception\Call\UnexpectedCallException: Method call:
  - hasAccess("edit", "my subject")
on Double\Sonata\AdminBundle\Admin\AbstractAdmin\P7 was not expected, expected calls were:
  - isChild()
  - getMenuFactory()
  - trans(exact("Dashboard"), exact([]), exact("SonataAdminBundle"))
  - trans(exact("My class"), exact([]), exact(null))
  - getRouteGenerator()
  - hasRoute(exact("list"))
  - hasRoute(exact("edit"))
  - isGranted(exact("LIST"))
  - isGranted(exact("EDIT"))
  - generateUrl(exact("list"))
  - generateUrl(exact("edit"), exact(["id" => "my-object"]))
  - getCurrentChildAdmin()
  - getIdParameter()
  - getRequest()
  - hasSubject()
  - getSubject()
  - toString(exact("my subject"))
  - getLabelTranslatorStrategy()
  - getClassnameLabel()

/media/cassiano/cassianosonataadim/SonataAdminBundle/Admin/BreadcrumbsBuilder.php:112
/media/cassiano/cassianosonataadim/SonataAdminBundle/Admin/BreadcrumbsBuilder.php:60
/media/cassiano/cassianosonataadim/SonataAdminBundle/Admin/BreadcrumbsBuilder.php:57
/media/cassiano/cassianosonataadim/SonataAdminBundle/Tests/Admin/BreadcrumbsBuilderTest.php:418

--

There was 1 failure:

1) Sonata\AdminBundle\Tests\Admin\BreadcrumbsBuilderTest::testGetBreadcrumbs
Sonata\AdminBundle\Admin\Pool::getContainer() was not expected to be called more than once.

/media/cassiano/cassianosonataadim/SonataAdminBundle/Admin/AbstractAdmin.php:2020
/media/cassiano/cassianosonataadim/SonataAdminBundle/Admin/AbstractAdmin.php:1979
/media/cassiano/cassianosonataadim/SonataAdminBundle/Tests/Admin/BreadcrumbsBuilderTest.php:213

FAILURES!
Tests: 7, Assertions: 5, Errors: 1, Failures: 1.

Legacy deprecation notices (12)
```

New errors! \o/

greg0ire commented:
You can replace `$this->once()` with `$this->any()`, or if you know how many times it is supposed to be called, `$this->exactly(42)`

cassianotartari commented:
Done. Now I have only ``hasAccess("edit", "my subject")`` above error.

greg0ire commented:
In `testUnitChildGetBreadCrumbs`, you have to mock the call to `hasAccess`. Here, I did not use the traditional mocking system but something more simple called "Prophecy". To do this mock, you have to write this `$admin->hasAccess('edit', 'my subject')->shouldBeCalled()`

cassianotartari commented:
Done. Almost knowing something.

New error:
```bash
$ phpunit Tests/Admin/BreadcrumbsBuilderTest.php 
PHPUnit 4.7.6 by Sebastian Bergmann and contributors.

..E....

Time: 157 ms, Memory: 15.00Mb

There was 1 error:

1) Sonata\AdminBundle\Tests\Admin\BreadcrumbsBuilderTest::testUnitChildGetBreadCrumbs
Prophecy\Exception\Call\UnexpectedCallException: Method call:
  - addChild("My subject", ["uri" => null])
on Double\ItemInterface\P3 was not expected, expected calls were:
  - getParent()
  - addChild(exact("My subject"), exact(["uri" => "/myadmin/my-object"]))

/media/cassiano/cassianosonataadim/SonataAdminBundle/Admin/BreadcrumbsBuilder.php:116
/media/cassiano/cassianosonataadim/SonataAdminBundle/Admin/BreadcrumbsBuilder.php:60
/media/cassiano/cassianosonataadim/SonataAdminBundle/Admin/BreadcrumbsBuilder.php:57
/media/cassiano/cassianosonataadim/SonataAdminBundle/Tests/Admin/BreadcrumbsBuilderTest.php:406

FAILURES!
Tests: 7, Assertions: 8, Errors: 1.

Legacy deprecation notices (16)
```

greg0ire commented:
Sorry, the call you have to make should actually grant access : `$admin->hasAccess('edit', 'my subject')->willReturn(true)->shouldBeCalled()`

cassianotartari commented:
CAN'T BELIEVE! 
```bash
$ phpunit Tests/Admin/BreadcrumbsBuilderTest.php 
PHPUnit 4.7.6 by Sebastian Bergmann and contributors.

.......

Time: 159 ms, Memory: 15.00Mb

OK (7 tests, 15 assertions)

Legacy deprecation notices (16)
```

greg0ire commented:
Great! Now run `make cs` to fix the style, `git commit --amend` and `git push -f`

cassianotartari commented:
I hope one day could understand better all this PR stuff, don't bother so much you guys and be a better contributor to Sonata Project. Thanks for the patience.

greg0ire commented:
> I hope one day could understand better all this PR stuff, don't bother so much you guys and be a better contributor to Sonata Project. Thanks for the patience.

Well thanks for not giving up! I think you understand pretty well already, and you overcame several hard things, like the rebase. All there is left now is adding more tests for the extension and the breadcrumbs builder itself, and you will be done!

cassianotartari commented:
I don't understand something, I did:

1. ``make cs``
2. ``git commit --amend``
3. ``git push -f``

But Netbeans was still showing that the files wasn't committed, so I run commit under UI interface of Netbeans and on command line ``git rebase --interactive origin/3.x`` with ``f`` lines plus ``git push --force``

Why 1,2,3 steps didn't worked?

greg0ire commented:
Oh sorry, You have to use `git add -p` between step 1 and 2

cassianotartari commented:
>Well thanks for not giving up! I think you understand pretty well already, and you overcame several hard things, like the rebase.

I will save this PR to future consulting. You passed a lot of knowledge here. Maybe could be nice have in the docs of Sonata how to fork and contribute to the project, some "rules" like the one commit...

>All there is left now is adding more tests for the extension and the breadcrumbs builder itself, and you will be done!

What are the next steps? Last commit I've corrected StyleCI under ``Sonata\AdminBundle\DependencyInjection\Compiler\AddDependencyCallsCompilerPass``. Someone added an extra space.

greg0ire commented:
There is the CONTRIBUTING.md , it explains a lot of things too, make sure you read it.

> Last commit I've corrected StyleCI under Sonata\AdminBundle\DependencyInjection\Compiler\AddDependencyCallsCompilerPass. Someone added an extra space. 

That's strange, but ok… there are no tests for the extension class yet (WTF?), so I won't ask you to write them. Next and final step IMO is : modify `testUnitBuildBreadcrumbs` line 405 to test your option with 'show', and see if it behaves normally.

cassianotartari commented:
> modify testUnitBuildBreadcrumbs line 405 to test your option with 'show', and see if it behaves normally.

How? Changing ``$action`` to ``show``? Somehow I need to create an admin with childadmin? :weary:

greg0ire commented:
> How?

Remember when you added a constructor? Well know is time to use it! Try `$breadcrumbsBuilder = new BreadcrumbsBuilder('use the option you invented here')`, and see how it goes (you will probably need to slightly change the tests, confirming that the show route is called, and not the edit route.

cassianotartari commented:
Should I change line 317 that has ``$breadcrumbsBuilder = new BreadcrumbsBuilder();`` to ``$breadcrumbsBuilder = new BreadcrumbsBuilder(array('child_admin_route' => 'show'));`` and add ``$admin->hasAccess('show', 'my subject')->willReturn(true)->shouldBeCalled();`` plus ``$admin->hasRoute('show')->willReturn(true);`` or should I create a entire test? Am I close?

Doing this it throws:
```
There was 1 error:

1) Sonata\AdminBundle\Tests\Admin\BreadcrumbsBuilderTest::testUnitChildGetBreadCrumbs
Prophecy\Exception\Call\UnexpectedCallException: Method call:
  - generateUrl("show", ["id" => "my-object"])
```
Is there any problem to generate show route adding ``id`` get parameter? It will not have effects, just visual, I think.

greg0ire commented:
> Am I close?

Yes, you should change that, it shows that your code works. It would be great to test both cases if you feel up to it. You can achieve this by adding more parameters to the test, like, say `$options` and `$expectedChildRoute`, and change the data provider accordingly.

> Is there any problem to generate show route adding id get parameter? It will not have effects, just visual, I think.

I don't get it… shouldn't the show view have an id?

cassianotartari commented:
> Yes, you should change that, it shows that your code works. It would be great to test both cases if you feel up to it. You can achieve this by adding more parameters to the test, like, say $options and $expectedChildRoute, and change the data provider accordingly.

Cool, I'll give a shot.

cassianotartari commented:
I changed constructor to ``$breadcrumbsBuilder = new BreadcrumbsBuilder(array('child_admin_route' => 'show'));`` as I said.

Added:
```php
$admin->hasAccess('show', 'my subject')->willReturn(true)->shouldBeCalled();
$admin->hasRoute('show')->willReturn(true);
$admin->generateUrl('show', array('id' => 'my-object'))->willReturn('/myadmin/my-object');
```

Removed:
```
$admin->hasAccess('edit', 'my subject')->willReturn(true)->shouldBeCalled();
```

And test is **OK**.

How should I do others tests?

The actual code is:
```php
    public function testUnitChildGetBreadCrumbs()
    {
        $menu = $this->prophesize('Knp\Menu\ItemInterface');
        $menu->getParent()->willReturn(null);

        $dashboardMenu = $this->prophesize('Knp\Menu\ItemInterface');
        $dashboardMenu->getParent()->willReturn($menu);

        $adminListMenu = $this->prophesize('Knp\Menu\ItemInterface');
        $adminListMenu->getParent()->willReturn($dashboardMenu);

        $adminSubjectMenu = $this->prophesize('Knp\Menu\ItemInterface');
        $adminSubjectMenu->getParent()->willReturn($adminListMenu);

        $childMenu = $this->prophesize('Knp\Menu\ItemInterface');
        $childMenu->getParent()->willReturn($adminSubjectMenu);

        $leafMenu = $this->prophesize('Knp\Menu\ItemInterface');
        $leafMenu->getParent()->willReturn($childMenu);

        $action = 'my_action';
        $breadcrumbsBuilder = new BreadcrumbsBuilder(array('child_admin_route' => 'show'));
        $admin = $this->prophesize('Sonata\AdminBundle\Admin\AbstractAdmin');
        $admin->isChild()->willReturn(false);

        $menuFactory = $this->prophesize('Knp\Menu\MenuFactory');
        $menuFactory->createItem('root')->willReturn($menu);
        $admin->getMenuFactory()->willReturn($menuFactory);
        $labelTranslatorStrategy = $this->prophesize(
            'Sonata\AdminBundle\Translator\LabelTranslatorStrategyInterface'
        );
        $labelTranslatorStrategy->getLabel(
            'dashboard',
            'breadcrumb',
            'link'
        )->willReturn('Dashboard');
        $admin->trans('Dashboard', array(), 'SonataAdminBundle')->willReturn(
            'Tableau de bord'
        );

        $routeGenerator = $this->prophesize('Sonata\AdminBundle\Route\RouteGeneratorInterface');
        $routeGenerator->generate('sonata_admin_dashboard')->willReturn('/dashboard');

        $admin->getRouteGenerator()->willReturn($routeGenerator->reveal());
        $menu->addChild('Tableau de bord', array('uri' => '/dashboard'))->willReturn(
            $dashboardMenu->reveal()
        );
        $labelTranslatorStrategy->getLabel(
            'my_class_name_list',
            'breadcrumb',
            'link'
        )->willReturn('My class');
        $labelTranslatorStrategy->getLabel(
            'my_child_class_name_list',
            'breadcrumb',
            'link'
        )->willReturn('My child class');
        $childAdmin = $this->prophesize('Sonata\AdminBundle\Admin\AbstractAdmin');
        $childAdmin->isChild()->willReturn(true);
        $childAdmin->getParent()->willReturn($admin->reveal());
        $childAdmin->getLabelTranslatorStrategy()
            ->shouldBeCalled()
            ->willReturn($labelTranslatorStrategy->reveal());
        $childAdmin->getClassnameLabel()->willReturn('my_child_class_name');
        $childAdmin->hasRoute('list')->willReturn(true);
        $childAdmin->isGranted('LIST')->willReturn(true);
        $childAdmin->generateUrl('list')->willReturn('/myadmin/my-object/mychildadmin/list');
        $childAdmin->trans('My child class', array(), null)->willReturn('Ma classe fille');
        $childAdmin->getCurrentChildAdmin()->willReturn(null);
        $childAdmin->hasSubject()->willReturn(true);
        $childAdmin->getSubject()->willReturn('my subject');
        $childAdmin->toString('my subject')->willReturn('My subject');

        $admin->hasAccess('show', 'my subject')->willReturn(true)->shouldBeCalled();
        $admin->hasRoute('show')->willReturn(true);
        $admin->generateUrl('show', array('id' => 'my-object'))->willReturn('/myadmin/my-object');

        $admin->trans('My class', array(), null)->willReturn('Ma classe');
        $admin->hasRoute('list')->willReturn(true);
        $admin->isGranted('LIST')->willReturn(true);
        $admin->generateUrl('list')->willReturn('/myadmin/list');
        $admin->getCurrentChildAdmin()->willReturn($childAdmin->reveal());
        $request = $this->prophesize('Symfony\Component\HttpFoundation\Request');
        $request->get('slug')->willReturn('my-object');

        $admin->getIdParameter()->willReturn('slug');
        $admin->hasRoute('edit')->willReturn(true);
        $admin->isGranted('EDIT')->willReturn(true);
        $admin->generateUrl('edit', array('id' => 'my-object'))->willReturn('/myadmin/my-object');
        $admin->getRequest()->willReturn($request->reveal());
        $admin->hasSubject()->willReturn(true);
        $admin->getSubject()->willReturn('my subject');
        $admin->toString('my subject')->willReturn('My subject');
        $admin->getLabelTranslatorStrategy()->willReturn(
            $labelTranslatorStrategy->reveal()
        );
        $admin->getClassnameLabel()->willReturn('my_class_name');

        $dashboardMenu->addChild('Ma classe', array(
            'uri' => '/myadmin/list',
        ))->shouldBeCalled()->willReturn($adminListMenu->reveal());

        $adminListMenu->addChild('My subject', array(
            'uri' => '/myadmin/my-object',
        ))->shouldBeCalled()->willReturn($adminSubjectMenu->reveal());

        $adminSubjectMenu->addChild('Ma classe fille', array(
            'uri' => '/myadmin/my-object/mychildadmin/list',
        ))->shouldBeCalled()->willReturn($childMenu->reveal());

        $childMenu->addChild('My subject')
            ->shouldBeCalled()->willReturn($leafMenu->reveal());

        $breadcrumbs = $breadcrumbsBuilder->getBreadcrumbs($childAdmin->reveal(), $action);
        $this->assertCount(5, $breadcrumbs);
    }
```

greg0ire commented:
> How should I do others tests?

Not sure you really need to change the other tests: you have covered the new case, and the ancient case should be covered by the test you just pasted. So IMO you're done.

greg0ire commented:
@sonata-project/contributors , can you give this a (final?) review?

OskarStark commented:
please correct the changelog in the PR header

cassianotartari commented:
> please correct the changelog in the PR header

Done. Is it ok like this?

greg0ire commented:
I think you can remove the `Changed` part, because it will end in a big list, and will definitely look weird ;)

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

cassianotartari commented:
> Could you please rebase your PR and fix merge conflicts?

This means just
```bash
git rebase --interactive origin/3.x
git push -f
```
?

greg0ire commented:
1. get latest updates : `git fetch --all`
2. start the rebase `git rebase origin/3.x`
3. boom! conflict! fix it with `git mergetool`
4. continue the rebase with `git rebase --continue`, and go back to 3. each time you have a conflict
2. `git push -f`

greg0ire commented:
ping @sonata-project/contributors , this needs a review.

OskarStark commented on DependencyInjection/Configuration.php:
```diff
@@ -73,6 +73,12 @@ public function getConfigTreeBuilder()
+                ->arrayNode('breadcrumbs')
+                    ->addDefaultsIfNotSet()
+                    ->children()
+                        ->scalarNode('child_admin_route')->defaultValue('edit')->info('Change the default route used to generate the link to the parent object, when in a child admin')->end()
```
please linebreak

OskarStark commented:
Review done

OskarStark commented:
minor i guess?

greg0ire commented:
> minor i guess?

Yup

cassianotartari commented:
Linebreak like this?

greg0ire commented on DependencyInjection/Configuration.php:
```diff
@@ -73,6 +73,13 @@ public function getConfigTreeBuilder()
+                    ->addDefaultsIfNotSet()
+                    ->children()
+                        ->scalarNode('child_admin_route')->defaultValue('edit')->info('Change the default route used to generate the link '
+                                .'to the parent object, when in a child admin')->end()
```
you can simply do 

```
->scalarNode('child_admin_route')
    ->defaultValue('edit')
    ->info('Change the default route used to generate the link to the parent object, when in a child admin')
->end()
```

cassianotartari replied:
:grin: ok... hold on

greg0ire commented on Tests/Admin/BreadcrumbsBuilderTest.php.orig:
```diff
@@ -0,0 +1,551 @@
+
+        $breadcrumbsBuilder->buildBreadCrumbs($admin->reveal(), $action);
+    }
+}
```
You are not supposed to commit this file

cassianotartari replied:
I don't get it, why?

greg0ire replied:
Nope. Look closely at its name, it's a `.orig` file. This kind of file appears when you run into a git conflict, and is here in case you want to know what the file looked like before the conflict.

greg0ire commented:
Rebasing should fix the build, see #4079

OskarStark commented:
@greg0ire i assigned you to this PR; because you and @cassianotartari already wrote a book here 😄

greg0ire commented:
I already gave my LGTM, this needs another one

OskarStark commented:
Thank you @cassianotartari 👍

greg0ire commented:
And congrats! You did it!

cassianotartari commented:
The soap opera is over! and it's Friday \o/ :beers:

Thank you so much for the patience!
